modules/social_features/social_group/src/Plugin/Field/FieldWidget/SocialGroupSelectorWidget.php

```diff
@@ -111,79 +112,103 @@ public static function create(ContainerInterface $container, array $configuratio
-   *
-   * @param \Drupal\Core\Entity\FieldableEntityInterface $entity
-   *   The entity for which to return options.
-   *
-   * @return array
-   *   The array of options for the widget.
+   * Gets a list of supported entity types.
+   */
+  protected function types(): array {
+    return ['node'];
+  }
+
+  /**
+   * {@inheritdoc}
    */
   protected function getOptions(FieldableEntityInterface $entity) {
-    // Must be a node.
-    if ($entity->getEntityTypeId() !== 'node') {
-      // We only handle nodes. When using this widget on other content types,
-      // we simply return the normal options.
+    if (!in_array($type = $entity->getEntityTypeId(), $this->types())) {
       return parent::getOptions($entity);
     }
 
-    // Get the bundle fron the node.
-    $entity_type = $entity->bundle();
-
     /** @var \Drupal\user\Entity\User $account */
     $account = $this->userManager->load($this->currentUser->id());
 
+    $groups_admin = $account->hasPermission('manage all groups');
+
     // If the user can administer content and groups, we allow them to
-    // override this. Otherwise we stick to the original owner.
-    if (!$account->hasPermission('administer nodes') && !$account->hasPermission('manage all groups')) {
-      $account = $entity->getOwner();
+    // override this. Otherwise, we stick to the original owner.
+    if ($entity instanceof EntityOwnerInterface && !$groups_admin) {
+      if ($type === 'node') {
+        $permission = 'administer nodes';
+      }
+      else {
+        $definition = $this->entityTypeManager->getDefinition($type);
+
+        if ($definition !== NULL) {
+          $permission = $definition->getAdminPermission();
+        }
+      }
+
+      if (
+        empty($permission) ||
+        !is_string($permission) ||
+        !$account->hasPermission($permission)
+      ) {
+        $account = $entity->getOwner();
+      }
     }
 
     // Limit the settable options for the current user account.
-    $options = $this->fieldDefinition
+    $options_provider = $this->fieldDefinition
       ->getFieldStorageDefinition()
-      ->getOptionsProvider($this->column, $entity)
-      ->getSettableOptions($account);
-
-    // Check for each group type if the content type is installed.
-    foreach ($options as $key => $optgroup) {
-      // Groups are in the array below.
-      if (is_array($optgroup)) {
-        // Loop through the groups.
-        foreach ($optgroup as $gid => $title) {
+      ->getOptionsProvider($this->column, $entity);
+
+    if ($options_provider !== NULL) {
+      $options = $options_provider->getSettableOptions($account);
+
+      $storage = $this->entityTypeManager->getStorage('group');
+
+      // Check for each group type if the content type is installed.
+      foreach ($options as $key => $optgroup) {
+        // Groups are in the array below.
+        if (is_array($optgroup)) {
+          $group = $storage->load(array_keys($optgroup)[0]);
```

Comment:
Wouldn't it be better to use the [key](https://www.php.net/manual/en/function.key.php) function?